storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -206,4 +223,119 @@ public static Optional<CompletedTxn> updateProducers(ProducerStateManager produc
+            }
+
+            public void recordInvalidChecksums() {
+                allTopicsStats.invalidMessageCrcRecordsPerSec().mark();
+            }
+
+            public void recordNoKeyCompactedTopic() {
+                allTopicsStats.noKeyCompactedTopicRecordsPerSec().mark();
+            }
+        };
+    }
+
+    /**
+     * If the recordVersion is >= RecordVersion.V2, create a new LeaderEpochFileCache instance.
+     * Loading the epoch entries from the backing checkpoint file or the provided currentCache if not empty.
+     * Otherwise, the message format is considered incompatible and the existing LeaderEpoch file
+     * is deleted.
+     *
+     * @param dir                  The directory in which the log will reside
+     * @param topicPartition       The topic partition
+     * @param logDirFailureChannel The LogDirFailureChannel to asynchronously handle log dir failure
+     * @param recordVersion        The record version
+     * @param logPrefix            The logging prefix
+     * @param currentCache         The current LeaderEpochFileCache instance (if any)
+     * @param scheduler            The scheduler for executing asynchronous tasks
+     * @return The new LeaderEpochFileCache instance (if created), empty otherwise
+     */
+    public static Optional<LeaderEpochFileCache> maybeCreateLeaderEpochCache(File dir,
+                                                                             TopicPartition topicPartition,
+                                                                             LogDirFailureChannel logDirFailureChannel,
+                                                                             RecordVersion recordVersion,
+                                                                             String logPrefix,
+                                                                             Optional<LeaderEpochFileCache> currentCache,
+                                                                             Scheduler scheduler) throws IOException {
+        File leaderEpochFile = LeaderEpochCheckpointFile.newFile(dir);
+
+        if (recordVersion.precedes(RecordVersion.V2)) {
+            if (leaderEpochFile.exists()) {
+                LOG.warn(logPrefix + "Deleting non-empty leader epoch cache due to incompatible message format " + recordVersion);
+            }
+            Files.deleteIfExists(leaderEpochFile.toPath());
+            return Optional.empty();
+        } else {
+            LeaderEpochCheckpointFile checkpointFile = new LeaderEpochCheckpointFile(leaderEpochFile, logDirFailureChannel);
+            return Optional.of(currentCache.map(cache -> cache.withCheckpoint(checkpointFile))
+                    .orElse(new LeaderEpochFileCache(topicPartition, checkpointFile, scheduler)));
+        }
+    }
+
+    public static LogSegment createNewCleanedSegment(File dir, LogConfig logConfig, long baseOffset) throws IOException {
```

Comment:
Do we truly need these additional methods? Directly calling `LocalLog.createNewCleanedSegment` seems sufficient to me.

Reply:
According to https://docs.google.com/document/d/1dQJL4MCwqQJSPmZkVmVzshFZKuFy_bCPtubav4wBfHQ/edit?usp=sharing we wanted UnifiedLog to be the public API other components will use:
```
This class is external facing, and acts as the outer shell, meaning that the public API of this class will be used by other components such as LogManager, LogCleaner etc. and other components outside the kafka.log package.
```

Reply:
Thanks for the documentation. It's good to know! :)